dvc/objects/file.py

```diff
@@ -51,3 +53,26 @@ def check(self, odb):
         assert actual.name == self.hash_info.name
         if actual.value.split(".")[0] != self.hash_info.value.split(".")[0]:
             raise ObjectFormatError(f"{self} is corrupted")
+
+
+class StreamFile(HashFile):
+    def __init__(self, path_info, fs, stream):
+        from dvc.utils.stream import HashedStreamReader
+
+        super().__init__(path_info, fs, None)
+        self.stream = HashedStreamReader(stream)
+
+    def sync(self, **upload_args):  # pylint: disable=unused-argument
+        with open(os.devnull, "wb") as stream:
+            shutil.copyfileobj(self.stream, stream)
```

Comment:
Seems like the only difference between `StreamFile` and normal file staging is that we compute the hash not right away, but when `sync` is called, which is effectively the same as normal file staging, because we call `sync` right away in `_transfer_file`.

The only special case is `TransferredStreamFile` one, where we combine uploading to a temp location with on-the-fly hash computation. This makes me think that we need to get rid of `transfer()` in favor of `stage()` that will accept some kind of strategy flag. Like `stage(...., upload=True)`. What do you think?